DM-31614: draft of butler provenance technote

TallJimbo opened the pull request.

TallJimbo committed:
Describe current middleware provenance support.

TallJimbo committed:
Add outline for remainder of technote.

timj commented on index.rst:
```diff
@@ -50,6 +50,133 @@
+TODO:
+
+- Support getting QG given (all optional): input DatasetRef/DatasetType/Task/Quantum, output DatasetRef/DatasetType/Task/Quantum, collections all datasets or quanta must be in.
+- Need to think about whether QG is well-defined for all argument combinations, and what arguments really mean; "input" and "outputs" are only well-defined for linear graphs, but there are probably definitions with that limiting behavior we can assume.
```
We should consider starting to use provenance nomenclature for provenance (as used by IVOA provenance model which is derived from the W3C provenance model) to be explicit about our meaning. (https://www.ivoa.net/documents/ProvenanceDM/20200411/REC-ProvenanceDM-1.0-20200411.pdf)

TallJimbo replied:
I think I picked up the "predicted" vs "actual" terminology from @MichelleGower , who I think got it from some (other? same?) standardized set of provenance definitions.  I did make up "available" on my own, though.  I'll scan that link to see if anything equivalent jumps out, unless Michelle can save me some time by clarifying where "predicted" and "actual" come from.

MichelleGower replied:
I would have said @TallJimbo came up with the names.  I definitely remember talking about the difference between the two categories and whether LSST wanted to track both, but not the origins of the names.

TallJimbo replied:
The IVOA provenance model seems to map to our `QuantumGraph` as follows:

- a `quantum` is one kind of (here, the only relevant kind of) _Activity_;
- a `dataset` is one kind of (here, the only relevant kind of) _Entity_;
- a `quantum` _Activity_ _Used_ an `input` `dataset`;
- an `output` `dataset` _Entity_ _WasGeneratedBy_ a `quantum` _Activity_.

IVOA also has:
- a _WasDerivedFrom_ relationship between two _Entities_;
- a _WasInformedBy_ relationship between two _Activities_.

These could be useful in exporting "collapsed" versions of the QG in which we drop datasets or quanta and instead define edges between the one we didn't drop.

We currently have no concept that maps to IVOA's _agent_.  IVOA has no concepts that map to the `predicted`, `actual`, and `available` distinctions I propose here, but there are free-form _UsageDescription_ and _GeneratedDescription_ fields where we can put them.

So, overall, I don't see a clear place to start using IVOA terminology in our own systems, at least not yet; the terms we might have borrowed already have well-established names on our side that are either more specific (`quantum` vs _Activity_ or `dataset` vs. _Entity_) or have an adjective form that fits better where we use the term (`input` or `output`) than a past-tense verb (_Used_, _WasGeneratedBy_).

I'll include this in the text, and I think it absolutely does make sense to reference the IVOA terms in other documentation, including code documentation.

timj replied:
We will have to be able to export provenance in the W3C naming style so we should at least be cognizant of the WasGeneratedBy, WasDerivedFrom scheme.

TallJimbo committed:
Remove placeholder comments.

TallJimbo committed:
Proofreading edits.

TallJimbo committed:
Note functionality gap for (BPS) job-level provenance.

TallJimbo committed:
Fill out schema and provenance-recording sections.

TallJimbo committed:
Add section describing mapping to IVOA provenance.

TallJimbo commented:
I'm not familiar with that directive (do we need to update https://developer.lsst.io/restructuredtext/style.html#citations?).  Do you have an example?  (If not, I'm sure I can go look it up.)

Should it just be the same URL you linked to above?

timj commented:
ADS has an entry here: https://ui.adsabs.harvard.edu/abs/2020ivoa.spec.0411S/abstract

You can copy that into a `local.bib` file and then

```
:cite:`2020ivoa.spec.0411S`
```

in the text. Technically before releasing the document we are meant to transfer that bibentry to the lsst-texmf bibentry and then rerun `make refresh-bib` (the README.rst talks about that).

TallJimbo committed:
Add citation (with temporary local.bib) for IVOA provenance.

TallJimbo committed:
Reformat footnote as inline note.

Footnote was confusing given that the bibliography also wants to use
footnotes with the same numbers.

TallJimbo committed:
Add draft of provenance query interfaces section.

TallJimbo committed:
Fill out section on inexact reproduction.

TallJimbo committed:
Add anchors and fix up links.

TallJimbo committed:
Fill out section on addressing provenance WG recommendations.

TallJimbo committed:
Add section on implementation sequencing.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
 
-  To add images, add the image file (png, svg or jpeg preferred) to the
-  _static/ directory. The reST syntax for adding the image is
+Processing data with LSST middleware involves executing a directed acyclic graph (DAG) called a ``QuantumGraph``, where each node is either a ``Quantum``, a single execution of a single ``PipelineTask``, or a dataset managed by the ``Butler``.
```
Is the "or" part about node true?  Maybe it's just the way I walk the graph that I only see the Quanta (each which "includes" the TaskDef, input dataset refs(?) and output dataset refs)

TallJimbo replied:
From the perspective of how our Python interfaces work, you're right, in that the `networkx` graph that our `QuantumGraph` Python classes currently exposes just contains quanta, but I think of that as just one view of the full quantum-dataset graph.  I'm not positive we'll need to change that in the code (not by removing the quantum-only view, but by providing other views, too); I think it's a possibility, and more importantly here I find the quantum-dataset version of the graph more useful for conceptual discussions.

MichelleGower replied:
I am fine with the conceptual description and thought that might be what you were doing.  But I guess because of the wording it sounded like a description of the implementation.

TallJimbo replied:
I'm going to leave this one as-is; I think it's the right description for the rest of the document, and explaining how there are other ways to view the graph would just get in the way this early in the implementation.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
-  To add images, add the image file (png, svg or jpeg preferred) to the
-  _static/ directory. The reST syntax for adding the image is
+Processing data with LSST middleware involves executing a directed acyclic graph (DAG) called a ``QuantumGraph``, where each node is either a ``Quantum``, a single execution of a single ``PipelineTask``, or a dataset managed by the ``Butler``.
+The datasets produced by executing a ``QuantumGraph`` are written to a single ``RUN`` collection in the data repository, with any other collections that hold input datasets associated with that ``RUN`` collection by defining a new ``CHAINED`` collection that includes both inputs and outs.
```
Sentence is difficult to read.  "with any other collections that hold input datasets associated with that RUN collection"  followed by "by defining"?  I think the clause actually ends is "that hold input datasets", but needs commas or rewording.

TallJimbo replied:
Yeah, I think I forgot a word somewhere, and this works better as two sentences.  Will fix.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
 
-  .. figure:: /_static/filename.ext
-     :name: fig-label
+- Dataset types with empty data IDs can be used to store arbitrary information describing a ``RUN``, as long as it is acceptable to obtain that information via ``Butler.get`` (i.e. it is not needed as a constraint on future ``Registry`` database queries).
```
What is the "it" in "it is not needed"?  The data ID?  The arbitrary information?

TallJimbo replied:
The arbitrary information.  Will reword.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+  Any kind of execution harness (including but not limited to BPS implementations) can also store their own per-``RUN`` information by defining their own dataset types, without any need to modify lower-level middleware software.
 
-     Caption text.
+- Dataset types with data IDs that match those of a particular ``PipelineTask's`` quanta can be use to store arbitrary information describing a quantum.
```
can be use -> can be used

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+- The ``CHAINED`` collection provides a record of the input collections that were used.
 
-   Feel free to delete this instructional comment.
+- The ``QuantumGraph`` itself is a complete description of the processing that can be used to reproduce a ``RUN`` exactly, on any system with access to the same datasets and sufficiently similar hardware and software (with the latter verifiable via the per-``RUN`` software versions dataset).
```
Just to nitpick the word "exactly", the QuantumGraph doesn't (as of this point in the document at least) contain whether a Quantum was executed or not.  A previous run that failed due to an intermittent system failure would not be exactly repeated (i.e., hopefully the new run would run more Quanta and produce more outputs).  And after reading a little further, this is indeed pointed out.

TallJimbo replied:
I'll add a caveat for this.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+- The software versions we record may not extend far enough into OS- or container-level software to guarantee reproducibility.
 
-.. TODO: Delete the note below before merging new content to the master branch.
+- Because we do not assume that the complete set of relevant software packages can be enumerated, and instead rely (at least partially) on what Python packages are imported, running additional processing that brings in new imports against an existing output ``RUN`` collection requires us to update the versions dataset in place - an operation we otherwise prohibit in ``Butler`` (datasets are consi ...
```
Will continuing processing be allowed after telling Rucio about files?  Because "deleting the old one and writing a new one - with the same filename because filenames are planned to be used as Rucio DIDs, won't work.

TallJimbo replied:
Good point. I already regarded this as a problem that we needed to solve, and Rucio makes it a much more concrete one, because our current workaround will fail.  We need to either switch to gathering versions in a way that does not rely on imports, or write a different software-versions dataset for each task (something I propose later in this document), since for a single task the set of imports should be constant.

TallJimbo replied:
I've added this point to the text.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
 
-.. TODO: Delete the note below before merging new content to the master branch.
+- Because we do not assume that the complete set of relevant software packages can be enumerated, and instead rely (at least partially) on what Python packages are imported, running additional processing that brings in new imports against an existing output ``RUN`` collection requires us to update the versions dataset in place - an operation we otherwise prohibit in ``Butler`` (datasets are consi ...
+  This is fragile but hard to fix, unless we just prohibit that kind of ``RUN`` collection extension or do enumerate all relevant software somehow up front.
```
Can there not be multiple software versions saved for a RUN and each Quantum tells which software version it used?

TallJimbo replied:
We could do this if we put each one in a different dataset type, e.g. named after tasks.  This is my preferred solution so far.

Having multiple software versions for different e.g. submissions (which might include some of the same tasks) is blocked for now, because we don't have a dimension that would let them coexist in the same collection. There are some other middleware feature requests pending that might change that, e.g. [DM-33751](https://jira.lsstcorp.org/browse/DM-33751), but I'd also like to have a solution that doesn't depend on that.

TallJimbo replied:
I'm not addressing this question at this point in the text because I want this list to be about problems, not solutions.  The solution proposed in my last comment will be raised later in the text.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+  It also fails when a ``CHAINED`` collection includes a "merge" of two output ``RUN`` collections that have conflicting datasets (i.e. they have same dataset type and data ID); dataset lookup then uses the order in which those ``RUN`` collections appear in the ``CHAINED`` collection to resolve the conflict, but this may not correspond to what was actually used as an input, and in fact more than  ...
+
+- Batch execution with BPS may "cluster" multiple quanta into a single "job" to e.g. avoid overheads in acquiring compute nodes or allow local scratch space to be used for some intermediate datasets.
+  We currently have no way to store job-level provenance in a data repository, except by denormalizing it into per-quantum datasets.
```
or waiting until the end and storing it in a single run-level provenance dataset?

TallJimbo replied:
Yes, that could work, provided we can save it and merge it outside the data repository until we are ready to include it in the data repository.  But right now we don't really have anything that conceptually "closes off" a RUN such that you can't do more processing in it later (as long as you use the same software versions and configs), and it seems like this would do that.  Of course, we are edging that way anyway, at least in BPS, by discouraging `--extend-run` in QuantumGraph generation.

TallJimbo replied:
Noted in the text.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+
+``quantum_id``
+  The quantum that produced this dataset (may be null).
+  Because a quantum can be produced by at most one quantum, we do not need a separate join table for this relationship.
```
Was this supposed to be "Because a dataset can be produced by at most one quantum"?

TallJimbo replied:
Yes, will fix.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+  A value of ``true`` does not guarantee continued existence in any ``Datastore``, but this this flag is still expected to provide an important optimization in queries for datasets that must be present in ``Datastores``, especially since most other reasons for dataset nonexistence allow checks to be implemented more efficiently by policy shortcuts (e.g. "this ``Datastore`` cannot have this ``PVI` ...
+
+The new ``task``, ``quantum``, and ``quantum_tags_*`` tables behave analogously to the ``dataset_type``, ``dataset``, and ``dataset_tag_*`` tables.
+Like datasets, quanta may be associated with collections, are "owned" by exactly one RUN collection, and there may be only one quantum with a particular data ID and task in each collection.
```
Perhaps this is answered later.  What about restarting a run?  Currently the merge job would have happened so the quantum rows would exist in the database.  Then someone restarts the run using same RUN collection, say because intermittent system error.  Is this going to be a clobber previous provenance or save multiple versions of the provenance where certain queries only return the provenance that produced the real datasets?

TallJimbo replied:
Very good point; I don't have an answer for this, and I think that's the weakest part of the proposal (at least given the comments I've read so far).

One possible solution that doesn't involve any any qualitatively new butler functionality/concepts:

- The merge job transfers failed quanta and their output datasets (which should not have been used as inputs by anything) to a _different_ RUN collection from the one that we use to hold successful outputs.
- We create a new such RUN every time a merge job runs (maybe skipping it when there are no failures).

Does that seem viable to you?  If so, any thoughts on what we call those RUN collections?

TallJimbo replied:
I've added a new subsection ("Failures and retries") discussing this and added a new `quantum.original_id` column to help with that.  It's more or less the idea above - that we have to move the failures out of the way - with a lot more detail added and some attempt to connect the failures to any eventual successful quantum.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
-   The butler stores datasets in collections along with software versions and pipeline configuration, but it does not store any information about the datasets that were used to generate a specific dataset. This document will discuss the plan for integrating provenance into Butler, sufficient to be able to regenerate any given dataset. It will refer to the requirements for provenance gathering spe ...
+Links between quanta and their inputs datasets are stored in the ``quantum_inputs`` table.
+This is a standard many-to-many join table with one extension: the ``actually_used`` flag.
+This may be set to ``false`` by tasks to indicate that the input dataset was not used *at all*, i.e. running the quantum without the dataset would have no effect on the results.
```
I'm probably getting confused by terminology.  Is task here the generic representation of PipelineTask or the specific execution of a Quantum?  I'm assuming it is each individual Quantum execution would either set the flag to true or false based upon its own inputs, but task seems to switch between the two meanings in this document.

TallJimbo replied:
Yes, I do use "task" to mean both "the code that runs a quantum" and "the entry in a pipeline with which a quantum is associated", and those are not quite the same thing.  This usage is just "the code that runs a quantum".

TallJimbo replied:
Changed "tasks" to to "`PipelineTask` implementations" in the text.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+
+A dataset may not have ``quantum_input.actually_used`` if ``dataset.actually_produced`` is ``false``.
+Note that these states build on each other; we say e.g. "predicted *only*" when an input is not available (and by extension, not actual).
+
```
Is the entire QuantumGraph going to be written to these tables or only those Quanta that were attempted to be executed?

TallJimbo replied:
My original intent was that we would save a quantum if we got far enough to have `SingleQuantumExecutor` check whether its predicted inputs were available, so if it was skipped at that stage it would still be written, but if the workflow system never saw it as unblocked, it wouldn't be written.  That works well in terms of division of responsibilities between lower-level middleware and BPS, I think, especially initially, while it's still the user's or BPS's responsibility to store the `QuantumGraph` file used to drive execution.  When we get all the way to the last "integrate provenance with pipetask" bullet in the implementation section at the end of this document, then I think it gets more natural to write the full predicted graph.  But if this is important earlier I'm sure we can find a way to make it work.

TallJimbo replied:
I've now clarified this in the text, but in the later section on actually recording the provenance.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+.. image:: /_static/tables.png
+   :target: https://dbdiagram.io/d/61fff3cc85022f4ee5479e62
+   :alt: registry provenance tables
+
```
Replicating host could add up.  What about Quantum "exit status"?

MichelleGower replied:
And what is dimension_graph_id  in the task table?   (I'm assuming that tag_association_table probably means a table name similary to quantum_tags_XYZ something like task_<label>_XYZ?)

TallJimbo replied:
Exit status is a good idea; will add.  I assume we just want an integer there, to match the process exit code?  I'm wary of trying to design anything too sophisticated or fully-enumerated in a SQL schema, as I'm not confident in our ability to get it right up front.

Do you think it's worth having a separate host table with an autoincrement key to allow them to be reused?

TallJimbo replied:
Adding a `status` `int` field.  We can consider moving host strings to a separate table at implementation time.  Using some kind of compressed string representation here with a lookup table outside the database (at least for frequently-used hosts) may also be worth considering.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+
+This approach works best if it is reflected in the in-memory ``QuantumGraph`` data structure and the execution model; these special init quanta would be executed prior to the execution of any of their task's usual quanta, which would be handled naturally by considering all task InitInputs to also be regular inputs of the task's regular quanta.
+This is actually more consistent with how BPS already treats the init job as just another node in its derived graphs, but with one init node per task, rather than one init node for the whole submission.
+Having one init quantum per task hints at a solution to another problem: if we write a different software version dataset for each (per-task) init quantum, instead of one for the entire RUN, each can be handled as a regular, write-once dataset, instead of needing to simulate update-in-place behavior.
```
What does this mean for checking that software versions are consistent?  (i.e., the reason behind sometimes needing the --no-versions option)

MichelleGower replied:
I think this may solve "add whole new tasks to the end of my run" update-in-place problems, but does it solve restarting a run update-in-place problems?

timj replied:
I'm also wondering about having to give people some more tools to work out if the software versions differed at all for the run.

TallJimbo replied:
This does indeed make the "check that I used the same versions" logic depend on what you're running, so it's a lookup on collection+task rather than just collection.  Other than being different from what we do now, I don't really consider that a problem.

As for restarting and rerunning tasks for which some quanta have already been executed in the same RUN, I think this proposal works as intended: you are allowed to do so if and only if you haven't changed software versions for that task _at all_.  That's much safer than what we do right now, where one is allowed to import new packages in a reprocess attempt, and if that import has side-effects, the outputs could change arbitrarily.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+Instead of adding the ``dataset.actually_produced`` column to the ``dataset`` table, we could ignore predicted-only datasets entirely in provenance.
+This simplifies the schema and avoids the problem how to query only for datasets that were actually produced.
+These predictions will still appear in ``QuantumGraph`` objects prior to execution, so by dropping them we make it impossible to reconstruct from provenance the full graph that was e.g. submitted to a batch system - but reproducing ``QuantumGraph`` generation is already not a guarantee of the system, because it depends on inputs (e.g. collections) that change over time.
+We would retain the ability to exactly re-execute any particular quantum and reproduce all output datasets however, because by definition these datasets are those that cannot have affected the results of the processing that did occur.
```
Are there any algorithms that say always throw out one or more of their inputs? (e.g., the worst)?  If so, these kinds of algorithms would still need the predicted in order to get the same actual as before.

Also, are there any algorithms that will continue running if a predicted input just can't be read at that moment?  What will be keeping track that something else should have been an input?

TallJimbo replied:
> Are there any algorithms that say always throw out one or more of their inputs? (e.g., the worst)? If so, these kinds of algorithms would still need the predicted in order to get the same actual as before.

My preferred definition of "actually used" would ban this: an "actually used" dataset is any dataset whose existence at execution time affects the outputs at all.

> Also, are there any algorithms that will continue running if a predicted input just can't be read at that moment? What will be keeping track that something else should have been an input?

If I understood your question correctly, then this is something that is quite common for tasks that take multiple inputs and rare (possible, but I don't know of any concrete examples) otherwise.

But under this alternative proposal, we would indeed lose all information about what might else have been used as an input.  This is _arguably_ okay because it is (from the task's perspective) indistinguishable from that input not being included in the QG originally, e.g. because of some user data ID expression filter, so remembering what was predicted at QG gen time is _just_ remembering what was predicted at QG gen time, not something deeper about possible other inputs.

TallJimbo replied:
See my new comment about this whole section, just above this one: I'm taking a middle ground and dropping this information from the database while keeping it in provenance datasets (i.e. files).

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+
+Traversal would be terminated in a direction whenever a ``while`` predicate evaluates to `False` (without returning that quantum) or whenever the corresponding ``until`` predicate evaluates to `True` (which does return that quantum).
+
+This simple conceptual definition of the predicate may not be possible in practice for performance reasons; traversal actually involves database queries, and while we can perform some post-query filtering in Python, we want most of the filtering to happen in the database.
```
Could some things be simplified by saving the actual QuantumGraph in the registry as a dataset that can be loaded?

TallJimbo replied:
I think it's cleaner not to to save the QG as a dataset - its data ID dimensions are problematic (at least until [DM-33751](https://jira.lsstcorp.org/browse/DM-33751)) as long as we permit multiple QGs per RUN, and after we bring the provenance home, the QG can be fully reconstructed from what's in the `Registry`.  If we also want to store everything in dataset form, too, then I'd lean towards duplicating the "predicted quantum" information into the per-quantum output provenance datasets, and keeping those around.

In the long term, this may make more sense, if DM-33751 provides a good data ID answer and pipetask+BPS integration leads to something like the "QuantumDirectory" third-butler-component I floated a while back.  But right now the way to work it into the design is sufficiently unclear that I'd rather not rely on it for anything.

TallJimbo replied:
In the latest version, I'm essentially advocating saving the original quanta + execution information as their own _separate_ datasets.  That's much cleaner in the sense that those quanta have well-defined data IDs and RUNs, but the `QuantumGraph` itself does not.

Storing individual quanta in their own files is less efficient, but it's how we'll write them originally, and if we need to fix that, I can still wave my hands like we always do about custom datastores that can put multiple datasets in a single file (which might actually just be a `QuantumGraph` file).

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+
+The most straightforward way to reprocess with changes is to create a completely new ``QuantumGraph``.
+The software versions and task configuration stored in per-``RUN`` datasets can be combined with the special "init" quanta for each task to reconstruct the pipelines used in a ``RUN`` exactly, and of course other versions or configurations may be used instead as desired.
+The input collections and data ID expression also typically provided as input to the ``QuantumGraph`` generation algorithm are not directly saved in our provenance schema, however, because after execution we intentionally do not draw any distinctions between quanta that may have originated in different graphs as long as their outputs land in the same ``RUN`` (and hence have the same software vers ...
```
I agree that these are not provenance because they are not frozen and can return different datasets over time.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+IVOA recommends certain predefined values be used for those descriptions when they apply (e.g. "Calibration" as a "UsedDescription"), which could be identified by configuration that depends on the pipeline definition.
+
+IVOA also defines a "WasInformedBy" relationship between two Activities and a "WasDerivedFrom" relationship between two Entities.
+These may be useful in collapsed views of the ``QuantumGraph`` in which datasets or quanta are elided, but in our case would always be computed from the Activity-Entity/quantum-dataset relationships, rather than meaningful in their own right.
```
I'm not sure what "meaningful in their own right" means here.  You mean there won't be a direct/stored representation of these relationships or that they aren't useful?

TallJimbo replied:
Meaning that they don't provide any new information (which is closely related to "not stored", of course).

TallJimbo replied:
I've adjusted the text to clarify.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+Because a relationship can have at most one associated "GeneratedDescription" or "UsedDescription", we cannot use one set of description types for the role-like information and another for this categorization.
+It seems best to simply leave this out of the IVOA view of our provenance, and possibly limit the IVOA view to "actual" relationships, since it is not expected to play a role in actually reproducing our processing.
+
+We currently have no concept that maps to IVOA's "Agent" or any of its relationships.
```
We could use Agent to be the person who executed the QuantumGraph which would be helpful internally.  But if looking for points of contact to be used by external folks, then it would be a name/group, email for pipelines or releases.

TallJimbo replied:
I wonder if @timj or @gpdf have opinions on what the usual granularity or expectations are here.  Would it be weird if we didn't have Agents in our provenance schema?  Is "Agent='Rubin Observatory'" something we should use if we don't have something else?

timj replied:
Pinging @gpdf on the Agent issue.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+This is already implemented and in regular use in the ``faro`` package in particular, and is currently the only way that metrics derived from butler data are initially stored.
+It is arguable whether the butler datasets are considered the source of truth after upload to SQuaSH.
+
+- *[REC-MET-003] When lsst.verify.Job objects are exported, the exported object should include the needed information (run collection and dataId) to associate with the source of truth metric persisted with the Data Butler.*
```
I'm assuming "Job" is an overloaded term and has nothing to do with compute jobs in bps.

TallJimbo replied:
Correct.  I would be happy if we could get rid of `lsst.verify.Job`, in fact, if reconciling our data model with theirs means it's no longer needed.  But I have no idea if that will be the case.

TallJimbo replied:
Clarified in the text.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+
+Avoiding per-dataset or per-quantum communication with a central SQL database is absolutely critical for at-scale execution with our middleware, so the provenance described above will need to be saved to files at first and loaded into the database later.
+
+Most of the information we need to save is already included in the ``QuantumGraph`` produced prior to execution, especially if we include UUIDs for its predicted intermediate and output in the graph at construction.
```
If UUIDs are saved in the graph at construction, we'll need other mechanisms for later creating new UUIDs if allowing folks to use same QuantumGraph (or Quantum Provenance) to execute a new run.

timj replied:
Are we going to persist the quantum graph itself in the RUN collection?

TallJimbo replied:
> If UUIDs are saved in the graph at construction, we'll need other mechanisms for later creating new UUIDs if allowing folks to use same QuantumGraph (or Quantum Provenance) to execute a new run.

If we're talking about a totally new RUN, then I'd express this as an API that takes an existing QG and makes a new one with new UUIDs and a new RUN collection (we are going to have to start baking RUNs into QGs soon for quantum-backed butler anyway).

If we're talking about retrying failed quanta, I think I'd also like to express it as a way to make a new QG, by generating new UUIDs (but keeping the RUN the same) for quanta that were failed but still brought home - or at least that's what's consistent with my proposal in response to your earlier question about storing failed quanta vs. retries.  It depends on how we handle that problem.

> Are we going to persist the quantum graph itself in the RUN collection?

See similar question from @MichelleGower below.

TallJimbo replied:
With the new "Failures and retries" section I've added in response to a different @MichelleGower comment, I think the resolution to this is:

 - We shouldn't have to make new UUIDs if we're reusing a QuantumGraph while skipping quanta from it that were already successful.
 - Having a way to rewrite the UUIDs and/or RUNs in a graph consistently still seems useful, especially as a way to re-do processing that was nominally successful but actually had problems.

MichelleGower reviewed (commented):
Here are comments from my first read through.  I don't think I fully understood everything enough to know whether I have any questions from the bps implementation viewpoint.  

I do think there needs to be a detailed discussion about when will rerunning exactly the same thing be allowed or not allowed (in light of when files are merged into central repo and more frozen Rucio) and whether this design saves all the information folks may want (especially in terms of previous attempts).   And whether one solution works everywhere or development/test runs need to have exceptions (e.g., clobber).

andy-slac commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+
+``quantum_id``
+  The quantum that produced this dataset (may be null).
+  Because a quantum can be produced by at most one quantum, we do not need a separate join table for this relationship.
```
It is true that one-to-many association does not need a separate table, but there may be reasons to keep quantum-dataset association in separate from dataset. One reason is to avoid schema migration for dataset table. Probably more general reason is to keep provenance implementation more separated from the rest of registry schema, e.g. to allow alternative implementations that do not need columns in the dataset table.

TallJimbo replied:
Your argument on on keeping the provenance implementation separate is a pretty compelling one - much of the reason the original quantum-provenance schema in Registry was removed during the transition to the "managers" classes was that the quantum manager class ended up having a circular dependency with the dataset one, due to exactly this.

It also avoids a circular FK relationship if we did want to add `dataset_id` columns to the `quantum` table for specific important provenance datasets.  I'm still not inclined to do that yet, but it's good to have that flexibility.

So I'll go make a `quantum_output` table with `dataset_id` (PK) and `quantum_id` columns to hold this instead.

andy-slac commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+  A dataset for which ``actually_produced`` is ``false`` should never appear in *any* ``Datastore``, and the dataset record exists only because the dataset was predicted to exist when a ``QuantumGraph`` was generated.
+  A value of ``true`` does not guarantee continued existence in any ``Datastore``, but this this flag is still expected to provide an important optimization in queries for datasets that must be present in ``Datastores``, especially since most other reasons for dataset nonexistence allow checks to be implemented more efficiently by policy shortcuts (e.g. "this ``Datastore`` cannot have this ``PVI` ...
+
+The new ``task``, ``quantum``, and ``quantum_tags_*`` tables behave analogously to the ``dataset_type``, ``dataset``, and ``dataset_tag_*`` tables.
```
`quantum_tags_*` worries me due its complexity and all management overhead. Do you think it may be possible to avoid having it completely and only use dataset-quantum associations for searching/navigation? I guess it depends on what kind of queries do we need to support. If we want to search quanta by exact DataId then it may be possible to encode DataId as a single column (JSON or something). Searching by inexact DataId will be hard to do, but maybe we don't need it? Alternative may be not to search by quantum DataId, but search for dataset(s) and navigate to quanta?

andy-slac replied:
Even crazier idea - could we store the whole per-quantum provenance as a regular dataset in registry/datastore. We'll lose all database niceties, but we will get more freedom in return. Querying performance is probably an issue with this approach, but we could do few smart things, and I'm not sure we care much about performance for provenance?

TallJimbo replied:
I don't see a way to avoid having at least the quantum-data linkage in the database; I think the best we could do without that would be to store a per-`RUN` QuantumGraph dataset, which would make basic provenance queries `O(size of RUN)`, and it raises all kinds of thorny issues about adding new quanta to an existing RUN.  And I think we want a quantum table for queries like "find me all the quanta that took longer than X minutes" or "find me all the quanta that ran on this bad node".

Whether we need data ID query support for quanta is less clear, and I think you're probably right that there we could get away with querying datasets by data ID and then going from dataset to quanta.  And if we didn't already have the analogous machinery for datasets in place, I'd completely agree that adding tags tables for quanta would be unnecessary complexity.  But given that we _do_ have that machinery, I think it is a nice _overall_ simplification both conceptually and in the code to just replicate that approach for quanta, rather that devising something new with different collection-membership concepts/constraints and data ID storage/querying.

One possibility in between might be to make it so a quantum "is a" dataset in the database, such that there is a row for each quantum in the `dataset` table, corresponding to some special per-task real dataset type for extended quantum information, since I think we want that anyway.  We would also still have `quantum` and `task` tables, but they would use `dataset_id` and `dataset_type_id` as their primary keys instead of having their own, with a subset of the values in `dataset` and `dataset_type`.  And then we could directly use the real `dataset_tag_*` tables to map quanta to data IDs and collections.  But to me this feels like abusing inheritance for implementation when it doesn't fit in terms of interface, especially if we map that inheritance-like structure to Python inheritance (and if we didn't, that inconsistency might bring its own problems).

So I'm still leaning more towards what I proposed here - we would want to refactor some of the manager code so we could reuse it, but once we've done that, I think this still is the best solution in terms of long-term, overall complexity.

TallJimbo replied:
I've come around on that last idea to address this concern: we're going to have a special per-quantum dataset anyway, so let's give it the same UUID as the quantum (we don't have to, but we might as well), and then just use the dataset's data ID and collections as the quantum's, with no limitations on what we can do with queries (just an extra 1-1 join in their implementation).

As a result, I've removed the `quantum_tags_XYZ` tables, and updated the text to discuss this.

andy-slac commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+Interfaces for querying quantum provenance
+------------------------------------------
+
+Given the similarity between quanta and datasets in terms of table structure, a ``Registy.queryQuanta`` method analogous to ``queryDatasets`` provides a good starting point for provenance searches::
```
Do we have requirements on provenance query API/functionality? I'm worried that trying to provide a very generic API we make things harder for ourselves.

timj replied:
We effectively have a global requirement to support the IVOA ProvDAL standard (which effectively means a way of implementing the ProvDM standard), although the server supporting such a thing vs butler registry supporting it natively can be two separate things.

TallJimbo replied:
I am also worried about this being very open-ended.  But what wrote down was very much informed by what I was confident we could implement:
 - for `queryQuanta`, I'm confident because it's basically a clone of `queryDatasets`.
 - for `traverseQuanta`, I'm confident because the naive implementation is just loading all of the quanta in a run into a graph and then doing networkx things with real predicate callbacks, but the interface is constrained enough that we can optimize important use cases as they come up.

TallJimbo replied:
I'm leaving this as-is for now since I don't have a better alternative (and the reasons in my last comment).  I did add a note that it should be interpreted as an example of capabilities, not as a detailed interface specification.

andy-slac reviewed (commented):
Few questions from me for database side of things. I think I'd like to understand if there are other possible implementation options for schema/API.

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+Before moving on, it is worth pointing out a few smaller issues with the provenance that we do record:
 
-.. sectnum::
+- The software versions we record may not extend far enough into OS- or container-level software to guarantee reproducibility.
```
It helps a lot that we now put almost everything (including compilers) into the conda-env and do record the entire state. We don't seem to record every eups package that is available though -- only the ones that we've imported (and then only through python `__version__`).

TallJimbo replied:
I'll add a note here about the way we use conda minimizing our dependence on the OS and container software.  I don't think it fully resolves the problem (and I think there was a Provenance WG entry highlighting this).

I think of the the "only record what we import" problem as being part of the bullet below.  I do think that's bigger problem (it sounds like you agree), and the best solution I've come up with so far is to write a new package versions dataset for each task - I think I worked that into some text later in the document, though it may have been buried a little.

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+- The software versions we record may not extend far enough into OS- or container-level software to guarantee reproducibility.
 
-.. TODO: Delete the note below before merging new content to the master branch.
+- Because we do not assume that the complete set of relevant software packages can be enumerated, and instead rely (at least partially) on what Python packages are imported, running additional processing that brings in new imports against an existing output ``RUN`` collection requires us to update the versions dataset in place - an operation we otherwise prohibit in ``Butler`` (datasets are consi ...
```
Again, we record more than python packages.

TallJimbo replied:
I think that's consistent with what I wrote:

> rely (at least partially) on what Python packages are imported

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+Avoiding per-dataset or per-quantum communication with a central SQL database is absolutely critical for at-scale execution with our middleware, so the provenance described above will need to be saved to files at first and loaded into the database later.
+
+Most of the information we need to save is already included in the ``QuantumGraph`` produced prior to execution, especially if we include UUIDs for its predicted intermediate and output in the graph at construction.
+We are already planning to do this for other reasons, as described in the `"Quantum-backed butler" proposal in DMTN-177 <https://dmtn-177.lsst.io/#limited-quantum-backed-butler>`__.
```
Can we also add a 
```
:cite:`DMTN-177` 
```
here so that the references at the end give a quick view of relevant documents? (also the other tech notes referenced later).

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+
+Most arguments here are exactly the same as those to ``queryDatasets``,
+with the dataset type argument replaced by a label expression identify the tasks, and two new arguments to constrain the query on particular input or output datasets.
+Like ``queryDatasets``, the return type would be a lazy-evaluation iterable, with convenience methods for conversion to ``QuantumGraph`` instance; this type could also be returned by a new ``DataCoordinateQueryResults.findQuanta`` method to more directly find quanta from a data ID query (as the ``findDatasets`` does for datasets).
```
I've been assuming that we will need something that given a DatasetRef returns all the parent input DatasetRefs and Quantum. Another use case is "what raws went into this dataset".

TallJimbo replied:
Yup, those are both things that I expect `traverseQuanta` (below) to handle.  We can provide convenience code on top of that for important things, but the real work is all down in `traverseQuanta`.

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+Intentionally inexact reproduction
+----------------------------------
+
+Quantum-based provenance excels at exact reproduction of previous processing runs, but it can also be used - with some limitations - to re-run processing with intentional changes.
```
Would it be worth doing a worked example of how we would do a PVI regeneration based on provenance information?

TallJimbo replied:
Ah, that's the funny thing - I don't think the usual PVI regeneration is actually a use case for provenance (or at least not a direct one) because the way we would _recreate_ a PVI is not at all the way we _originally_ created it.

To be precise, in the original DRP pipelines, we'd do something like (oversimplifying a bit):
```
isr -> image characterization 1 -> fgcm/jointcal -> image characterization 2 -> reprocess
```
and the PVI would be output by that last step, which naturally aggregates all of the various characterizations/calibrations produced along the way while it performs final source measurements.

But when recreating the PVI, we'd do something like:
```
isr -> recreate
```
Because all of the other modifications to the pixels that we'd want to keep are stored long-term in much smaller datasets (backgrounds, updated masks), and the `recreate` task could just load and apply those.

Now, we _could_ actually generate QGs for `isr -> recreate` in advance, and save them as provenance without actually running them.  That's an intriguing idea but not the only way to do it (I'd originally hoped to discuss this in detail back in that image-cutout-service discussion, but we didn't have time).  And I think it's probably best considered out of scope for this document.

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+- *[REC-EXP-1] As planned, program details known to the scheduler (such as science programme and campaign name) should be captured by the Butler.*
+
+As long as these details are included in the FITS headers of the raw files, they can be configured to be capture by the butler and stored in the ``exposure`` dimension table.
+This is already the case for the fields used as examples in this recommendation.
```
Only science programme is stored in exposure record. I'm not sure what campaign name means. Science programme was meant to cover survey (WFD, DDF etc) or commissioning program (we don't have a convention for labeling those).

TallJimbo replied:
I can update the text accordingly, but it sounds to me like it's probably a case of a poor example from DMTN-185, not something actually missing?

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+- The association of group ID with visits also discourages using group ID for more flexible groups of non-science exposures, such as focus sweeps or flat sequences, or causes confusion about whether visits (or some new visit-like dimension) should be defined and used for these.
+
+It would be much better for a new snap-specific header key to be introduced and interpreted by the middleware, freeing up the group ID to have more flexible and ad-hoc definitions.
+We believe this change is already ticketed in the data acquisition systems, but the work is not yet complete.
```
DM-33602 is the header service side and that has been done. We will still need to do DM-30948 for the butler side.

TallJimbo replied:
Thanks; I'll clarify that in the text.

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+
+- *[REC-MET-005] Even if effort for implementation is not available in construction, we should develop a conceptual design for structured, semantically rich storage of metrics in the Butler.*
+
+We currently save metric measurements as individual JSON files, which is convenient for upload to SQuaSH but inconvenient for querying metric values via the butler.
```
I think it might be YAML files but that's really a very inefficient way of doing it and we should be using JSON (there is a native JSON serialization -- the YAML one is using unsafe loading).

TallJimbo replied:
I'll just say "YAML or JSON" here, so the text isn't wrong - I don't think the distinction matters for the purposes of the technote.

TallJimbo replied:
Updated to reflect not just YAML vs. JSON but the switch to multiple measurements per JSON dataset with `analysis_tools` as well.

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+- *[REC-MET-002] Any system that uses Butler data to derive metrics should persist them in the Butler provided that the metrics are associable with a Data ID.*
+
+This is already implemented and in regular use in the ``faro`` package in particular, and is currently the only way that metrics derived from butler data are initially stored.
+It is arguable whether the butler datasets are considered the source of truth after upload to SQuaSH.
```
Cite SQR-008 and maybe mention/cite DMTN-203.

TallJimbo replied:
Will do.

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+
+- *[REC-MET-003] When lsst.verify.Job objects are exported, the exported object should include the needed information (run collection and dataId) to associate with the source of truth metric persisted with the Data Butler.*
+
+It is at least unusual for butler datasets to store their own data ID and especially their own ``RUN`` collection internally.
```
We talked about allowing formatters to support an "augment metadata" method which would allow us to store the dataId and run in the file itself during write. You address the other part of that in 3.4.

TallJimbo replied:
True; I'll add a forward-reference here, saying that it's a possibility that it will become less unusual and better-supported.

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+That metadata should be passed through a new optional keyword argument to ``Butler.put``, and can be prepared by the execution system to include provenance information.
+As implied by the recommendation text, this could only be implemented in formatters that use data formats that can store flexible metadata, but this should be true in practice for any data format used for public data products (including FITS, Parquet, and JSON).
+
+To satisfy *[REC-FIL-1]*, all we need to store is the dataset's UUID, which (after implementing the design described in [:ref:`recording-provenance`]) will be available to the execution system for insertion into the provenance metadata when the dataset is first saved to disk.
```
Datastore already has the DatasetRef so no provenance system changes are needed to write the UUID into the file metadata. We would need to write the formatter support for it.

TallJimbo replied:
Good point; I was thinking ahead to storing the quantum UUID (I assume we won't be adding that to `DatasetRef`), but not needing to add plumbing in higher-level interfaces is a very good reason to just add the dataset UUID and stop there.

I'll rewrite this section to highlight that possibility and probably make it my new recommendation.

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+
+Low-level I/O for files written by the butler goes through the ``lsst.daf.butler.Formatter`` interface, which could be easily extended to include external ``dict``-like metadata that should be recorded in the file.
+That metadata should be passed through a new optional keyword argument to ``Butler.put``, and can be prepared by the execution system to include provenance information.
+As implied by the recommendation text, this could only be implemented in formatters that use data formats that can store flexible metadata, but this should be true in practice for any data format used for public data products (including FITS, Parquet, and JSON).
```
JSON is tricky isn't it? There is no metadata concept in JSON. At issue is whether the python object itself supports metadata extensions that will appear in the JSON.

TallJimbo replied:
Good point, though to some degree that's _also_ true of Parquet and especially FITS, where we store important parts of our in-memory objects in the metadata, and conceptually that is really quite distinct from flexible metadata that the formatter can write to freely, even if it uses the same mechanism and part of the file.

TallJimbo replied:
Added this caveat to the text.

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+This may be less useful than it seems or perhaps even slightly confusing in some cases, however, because:
+
+- as of the time a dataset is written, we can only reliably know the "available" inputs to the quantum; the task may *later* declare that only some of these inputs were actually used;
+- in a few cases (large gather-scatter sequence points in the graph), the number of inputs to the quantum may be large (in the case of FGCM, a full-survey sequence point, it will be enormous);
```
You could conceive of having per-dataset type control over this so for fgcm we would not try to persist all inputs. Write parameters can do this already.

TallJimbo replied:
Good idea.  I'm still inclined not to try to support this at all (especially after you pointed out how it is is to write just the dataset UUID), but that's what we ought to do if we do want this.

timj commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+
+These seem like dangerous assumptions, in that satisfying them probably either creates unwanted dependencies between software components (the execution system has to be configured to know about RSP service endpoints) or puts undesirable constraints on others.
+
+Injecting this metadata into files when they are retrieved seems much cleaner conceptually, but it may rule out simple and/or efficient approaches to data access that would otherwise be on the table.
```
We have pondered doing this type of file rewriting on get in order to support a request for HDF5 version of the dataset. Injecting FITS headers on get could work at the cost of performance (and checksums changing)

timj reviewed (approved):
Looks great. Thanks for the detail.

TallJimbo reviewed (commented):
Thanks for all the comments.  I believe I've responded to all of them, and I'll start work implementing the trivial changes shortly, and will resolve them as I go.  I'll wait a bit to start work on bigger changes in case there's more discussion.

TallJimbo commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+InitInput and InitOutput datasets
+"""""""""""""""""""""""""""""""""
```
I've upgraded this section to part of the main proposal rather than an "open question". I think it solves a lot of problems, and doesn't introduce any new ones.

TallJimbo commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+Having one init quantum per task hints at a solution to another problem: if we write a different software version dataset for each (per-task) init quantum, instead of one for the entire RUN, each can be handled as a regular, write-once dataset, instead of needing to simulate update-in-place behavior.
+
+No predicted-only datasets
+""""""""""""""""""""""""""
```
I've merged this variant proposal with a solution to @andy-slac 's concern about complexity introduced by a bunch of `quantum_tags_*` tables, and made that the main proposal.  Summary is:

- We have a special dataset for each quantum that contains the full information about an executed quantum (it's probably the file the Quantum-backed butler writes).  This dataset's data ID and collections serve as the data ID and collections of the quantum.
- The database representation of a quantum is a summary that includes all the linkage to actual datasets and metadata we expect to want for provenance queries.
- Predicted-only datasets are remembered by that special quantum dataset, but are not loaded into the database.  This saves us from having a ton of registry rows for datasets that never existed in any datastore, and for the most part couldn't ever exist (e.g. warps for visits that don't actually overlap their patches).

TallJimbo commented on index.rst:
```diff
@@ -1,61 +1,486 @@
+We would retain the ability to exactly re-execute any particular quantum and reproduce all output datasets however, because by definition these datasets are those that cannot have affected the results of the processing that did occur.
+
+Reverse-lookups for special datasets
+""""""""""""""""""""""""""""""""""""
```
I've just dropped this proposal modification from the text. Nobody seems fond of it, so it's just noise.

TallJimbo committed:
Address review comments.

This includes major changes to the schema with a greater reliance on
provenance information in datasets (not just Registry), a detailed
discussion for how to handle failures and retries, and numerous small
miscellaneous issues.

TallJimbo committed:
Update bibliographies.

TallJimbo committed:
Use :cite: instead of manual links for technotes.

TallJimbo committed:
Update section on DMTN-185 recommendations.

TallJimbo reviewed (commented):
I've just made a few big changes and a large number of smaller ones, and I think this is ready another look by all interested reviewers.  If you're up for going through again, please hit "Resolved" on any comments you're done with; I've done that for some particularly trivial ones, but didn't want to do it for most.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,499 @@
+
+Our proposal is to only load provenance into the ``Registry`` database if the execution system has at least attempted to run them.
+Quanta whose executions were blocked by failures in upstream quanta would not be included, even if present in the original ``QuantumGraph``.
+This leaves room for them to be added ny later submissions of the same graph (or subsets thereof) without any UUID conflicts.
```
typo "ny"

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,499 @@
+IVOA recommends certain predefined values be used for those descriptions when they apply (e.g. "Calibration" as a "UsedDescription"), which could be identified by configuration that depends on the pipeline definition.
+
+IVOA also defines a "WasInformedBy" relationship between two Activities and a "WasDerivedFrom" relationship between two Entities.
+These may be useful in collapsed views of the ``QuantumGraph`` in which datasets or quanta are elided, but in our case they ca always be computed from the Activity-Entity/quantum-dataset relationships, rather than being graphs we would store directly.
```
typo "ca"

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,499 @@
+      ...
+
+Traversal could proceed forward (in the same direction as execution) or backward (from outputs back to inputs) or both.
+The criteria for which quanta to traverse and return are encoded in the four predicate arguments, which are *conceptually* just boolean functions on a quantum::
```
I think I get the high level concept, but not the details.  Given a set of Quanta, is traverseQuanta restricted to same RUN collection?  Is limiting on RUN collection part of the TraversalPredicate?

TallJimbo replied:
Limiting on RUN collection would be part of the `TraversalPredicate`.

I've added this to the list of example predicates below to make this more clear.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,499 @@
+
+It is worth noting here that the ``Quantum`` and ``QuantumGraph`` objects returned here are not necessarily the same types as those used prior to execution; execution adds more information that we want the provenance system to be able to return.
+Whether to actually use different types involves a lot of classic software design tradeoffs involving inheritance and container classes, and resolving it is beyond the scope of this document.
+If we do use different types, one of the most important operations on the "executed" forms will of course be transformation to a "predicted" form for reproduction.
```
Is there a planned interface for querying dataset provenance?  For example:

- Post-processing a config was deemed to have "bad" value(s).  What runs or do get more precise what data products were produced using this bad config?  Similar question for a "bad" raw or calibration file.  (Doesn't have to be bad, it could be an interesting feature that want to see in different coadds, for example)
- Given a coadd, what raws (or other dataset type in between) affected the coadd.
- Want to make sure particular datasets aren't relevant to a data release before purging.

TallJimbo replied:
> Post-processing a config was deemed to have "bad" value(s).  What runs or do get more precise what data products were produced using this bad config?

I think the first step of this search has to be more on the campaign management / change control / software versions side.  If we git-control all configs (or, even better, only run configs that are part of released software versions), then we'd use git and similar tooling to find config "versions" that had those bad configurations (and most of the time, I'd imagine the answer looks like "bad before commit X, good after").  I don't think there's any way we can index the contents of the actual config datasets well enough to perform a search on those efficiently.

> Similar question for a "bad" raw or calibration file. (Doesn't have to be bad, it could be an interesting feature that want to see in different coadds, for example)

This one is much easier: it's `queryQuanta(..., with_inputs=bad_refs)`, followed by `traverseQuanta(..., forward_while: lambda quantum: True)` and then yielding datasets from those traversed quanta.

> Given a coadd, what raws (or other dataset type in between) affected the coadd.

Similar to the last one, but with `with_outputs=` instead of `with_inputs=`, and then traversing backward instead of forward until you see `raw`.

> Want to make sure particular datasets aren't relevant to a data release before purging.

If "relevant to a data release" actually means included in the data release, then I don't think this requires provenance, so I assume you mean "somewhere in the history of a data release product."  It's technically doable with `traverseQuanta`, but it'd need a really efficient (certainly parallel) implementation of that, or more likely a bunch of parallel calls that were willing to live with some duplicate work, since with multiple parallel calls we'd no longer be able to guarantee each quanta is only traversed once.  But I think this is a hard-to-scale problem no matter how you try to implement it; it's basically a traversal of the entire DR provenance graph with a test for "DatasetRef in set" on all inputs.

Anyhow, we may well want a high-level interface on some or all of these, but the set of all possible high-level interfaces currently feels so overwhelmingly large that I didn't want to call out any right now.  It may also be that rewriting the traversal interface to give the predicate (or some other callback that's a replacement for those predicates) more control over which edges to traverse would make this a lot more efficient.  I think that's worth thinking about when we get closer to implementation, and something worth mining other provenance-query systems for ideas on.  For now I've just added some hedges about variants on this we might explore.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,499 @@
+This leads to a conflict: running a failed quantum again would lead to two ``quantum`` rows with the same UUID, data ID, and ``RUN``, and analoguous conflicts for any output datasets that were written before the failure occurred or in spite of it (e.g. logs written by the execution system after it catches an exception).
+
+The possibility of immediate retries performed by the workflow system means we cannot assign the retry quantum and datasets new UUIDs, because this would prevent downstream quanta from being able to find their inputs.
+Instead, we have to assign new UUIDs (and a new RUN collection) to the failures, and move all datasets that were already written to permanent storage out of the way.
```
"move all datasets that were already written to permanent storage out of the way"
This seems time consuming as well as possibly problematic if we're talking about the retry starting after the merge has been done (e.g., when will Rucio be told about the files)

TallJimbo replied:
Agreed.  This is why I want the transfer job to do this with any failures it sees that haven't already been moved out of the way, before the likes of Rucio sees them.  Those moves are still going to be somewhat expensive, and dropping the failure artifacts when the failure mode is super obvious (like preemption) will probably be important, but I don't see any other option.

If we have to tell Rucio about the files before we get to the transfer job for other reasons, then this gets a lot messier, and I'm not sure what we can do besides bite the bullet and move them in Rucio.

MichelleGower commented on index.rst:
```diff
@@ -1,61 +1,499 @@
+
+The possibility of immediate retries performed by the workflow system means we cannot assign the retry quantum and datasets new UUIDs, because this would prevent downstream quanta from being able to find their inputs.
+Instead, we have to assign new UUIDs (and a new RUN collection) to the failures, and move all datasets that were already written to permanent storage out of the way.
+This could be done immediately after the failure occurs if we have a process able to do the work (i.e. the worker process fails gracefully, or some control process is able to take action), or it could be done only just before actually attempting a retry.
```
The safest would be before actually attempting a retry (e.g., preemption on the IDF).  Is everything in a recoverable state if something happens between moving some files and actually rerunning the Quanta?

TallJimbo replied:
Agree that it's safest to do right before attempting a retry, so we should always do it there, but I also think we always need to do it in the transfer job for cases that haven't been retried, to avoid putting non-permanent locations in (e.g.) Rucio and on-permanent UUIDs in the main registry when at all possible.

I think it's all recoverable if the logic responsible for this "move out of the way" operation is written carefully:
- write the updated UUIDs and their connections to a new Quantum file in a deterministic location (or at least a location we can quickly search for the existence of such a file);
- only move other artifacts to that location after it has been written.

MichelleGower reviewed (commented):
Some questions/comments after reading the updated document.

TallJimbo committed:
Address latest round of review comments.

andy-slac commented on index.rst:
```diff
@@ -1,61 +1,504 @@
+This dataset does not need to include all provenance information related to the quantum; the execution system can chose to write arbitrary additional information (e.g. logs) to different dataset types that are related to the quantum in the same way as its regular output datasets (these are still identifiable as special via their dataset types).
+The ``${TASK}_quantum`` dataset should not appear as a regular output of the quantum, and it should not be used as an input by any downstream task.
+The ``host``, ``status``, and ``timespan_*`` columns of the ``quantum`` table provide basic metadata about the execution that we consider likely to be search criteria in provenance queries.
+The ``original_id`` columns provides the original UUID for quanta whose executions failed, as described below in :ref:`failures_and_retries`.
```
columns -> column?

andy-slac commented on index.rst:
```diff
@@ -1,61 +1,504 @@
+This interface does not provide enough functionality for most provenance queries, however - it just finds all quanta matching certain criteria, regardless of their relationships - so it is best considered way to obtain a starting point.
+For those, we envision an operation that starts with a set of quanta and traverses the graph according to certain criteria, querying the database as necessary (perhaps once per task or dataset type) and returning matching quanta as it goes::
+
+  def traverseQuanta(
+      self,
+      initial: Iterable[Quantum],
+      forward_while: Optional[str | TraversalPredicate] = None,
+      forward_until: Optional[str | TraversalPredicate] = None,
+      backward_while: Optional[str | TraversalPredicate] = None,
+      backward_until: Optional[str | TraversalPredicate] = None,
+  ) -> Iterable[Quantum]:
+      ...
```
Given that the structure of a provenance graph should reflect the structure of the original `QuantumGraph`, would it make sense to have this functionality (only) in `QuantumGraph` (or some common interface of QuantumGraph and whatever can be constructed from `QuantumQueryResults`)?

TallJimbo replied:
I think it's a good idea to at least put something very similar on all of our QuantumGraph-like structures, and especially to allow `TraversalPredicates` that don't involve fields that are only present in already-executed quanta to work on as many data structures as possible.  I don't want to assume yet that the interfaces will be identical, because traversing an in-memory graph may need to be different in some ways from querying a graph that's still mostly just in the database.

andy-slac commented on index.rst:
```diff
@@ -1,61 +1,504 @@
+      def __call__(self, quantum: Quantum) -> bool:
+          ...
+
+Traversal would be terminated in a direction whenever a ``while`` predicate evaluates to `False` (without returning that quantum) or whenever the corresponding ``until`` predicate evaluates to `True` (which does return that quantum).
```
I'm not sure this is general enough to cover all cases. I can imagine that sometimes you need to see two nodes connected by a graph edge to decide whether you want to follow that edge. But then I'm not even sure what is the use case for this method.

TallJimbo replied:
That's possible, but I can't come up with any use cases for it right now.  I'm not opposed to extending the interface if it comes up.

I'm really just taking a guess at a good way to support lots of kinds of graph traversal here; the real test will be looking at whether various expected provenance queries can be mapped to it (and for the ones I've thought of, that's the case).  I haven't tried very hard to see if there's a similarly general but simpler interface, though.

andy-slac commented on index.rst:
```diff
@@ -1,61 +1,504 @@
+
+- *[REC-MET-003] When lsst.verify.Job objects are exported, the exported object should include the needed information (run collection and dataId) to associate with the source of truth metric persisted with the Data Butler.*
+
+At prsent it is unusual for butler datasets to store their own data ID and especially their own ``RUN`` collection internally, but this is something expect to do more in the future (see discussion on *REC-FIL-1* below).
```
typo: prsent

"this is something expect to do more" does not parse :slightly_smiling_face:

andy-slac reviewed (approved):
Looks good (does not mean I understood every detail :slightly_smiling_face:), few minor comments.

TallJimbo committed:
Fix typos and minor word omissions.

timj commented on index.rst:
```diff
@@ -1,61 +1,504 @@
+
+- The approach of using datasets to record arbitrary provenance for higher-level tooling is currently extensible only to the extent that those datasets can use existing storage classes.
+  New storage classes must currently be registered by editing configuration in the data repository or ``daf_butler`` manually.
+  This is a problem for much more than just provenance, however (it also limits pipeline extensibility), and is something we hope to resolve.
```
Butler does allow you to define your own storage classes anywhere and set an environment variable to tell butler where to find the new definitions. We've never really used that feature though.

TallJimbo replied:
Added some text to mention this.

TallJimbo reviewed (commented):
Forgot to hit the submit button earlier on this set of replies to @andy-slac .

TallJimbo committed:
Mention envvars as possible solution to storage class extensibility.

TallJimbo committed:
Delete "not yet published" note.
